### DISCUSSION TITLE
RMW Compression Heuristic for In-Progress Ransomware Attack?

### DISCUSSION THREAD

**Discussion by sempervictus** [openzfs/zfs#12558] [Ideas]
When "ransomware" encrypts file data in-place, blocks which used to contain compressible data suddenly contain rather uncompressible data. Since their intent is to encrypt as much data as possible, the compression ratio should drop like a brick as the process goes on. How difficult would it be to teach ZED or the like to generate an event when this is observed such that HIDS or other response mechanisms could evaluate what's going on and try to respond accordingly?

**Comment by IvanVolosyuk** [comment]
If daily snapshots are used, there is no immediate threat of losing long
term data. More likely you will notice sudden increase in storage used and
compress ratio of recent snapshots dropping to 1, which is actually normal
for media files.
I also read an article about SSD-level heuristic to detect this kind of
activity. Hopefully it will not detect normal ZFS operations as suspicious.
